neo/core/spiketrainlist.py

```diff
@@ -0,0 +1,294 @@
+            combined_spike_time_array = np.hstack(
+                (self._spike_time_array, other._spike_time_array))
+            combined_channel_id_array = np.hstack(
+                (self._channel_id_array, other._channel_id_array))
+            combined_channel_ids = set(list(self._all_channel_ids) + other._all_channel_ids)
+            if (len(combined_channel_ids) !=
+                    len(self._all_channel_ids) + len(other._all_channel_ids)):
+                raise ValueError("Duplicate channel ids, please rename channels before adding")
+            if in_place:
+                self._spike_time_array = combined_spike_time_array
+                self._channel_id_array = combined_channel_id_array
+                self._all_channel_ids = combined_channel_ids
+                self._items = None
+                return self
+            else:
+                return self.__class__.from_spike_time_array(
+                    combined_spike_time_array,
+                    combined_channel_id_array,
+                    combined_channel_ids,
+                    t_start=self._spiketrain_metadata['t_start'],
+                    t_stop=self._spiketrain_metadata['t_stop'])
+
+    def __add__(self, other):
+        """Return self + other"""
+        if isinstance(other, self.__class__):
+            return self._add_spiketrainlists(other)
+        elif other and is_spiketrain_or_proxy(other[0]):
+            return self._add_spiketrainlists(
+                self.__class__(items=other, segment=self.segment)
+            )
+        else:
+            if self._items is None:
+                self._spiketrains_from_array()
+            return self._items + other
+
+    def __iadd__(self, other):
+        """Return self"""
+        if isinstance(other, self.__class__):
+            return self._add_spiketrainlists(other, in_place=True)
+        elif other and is_spiketrain_or_proxy(other[0]):
+            for obj in other:
+                obj.segment = self.segment
+            if self._items is None:
+                self._spiketrains_from_array()
+            self._items.extend(other)
+            return self
+        else:
+            raise TypeError("Can only add a SpikeTrainList or a list of SpikeTrains in place")
+
+    def __radd__(self, other):
+        """Return other + self"""
+        if isinstance(other, self.__class__):
+            return other._add_spiketrainlists(self)
+        elif other and is_spiketrain_or_proxy(other[0]):
+            for obj in other:
+                obj.segment = self.segment
+            if self._items is None:
+                self._spiketrains_from_array()
+            self._items.extend(other)
+            return self
+        elif len(other) == 0:
+            return self
+        else:
+            if self._items is None:
+                self._spiketrains_from_array()
+            return other + self._items
+
+    def append(self, obj):
```

Comment:
If the inetrnal state is multiplexed then this implementation cost a lot.
Maybe we could add an option to transform other to spiketrainlist+remapp ids+concatenate.

Reply:
I think a lot of internal optimisations are possible, but they should be done in separate PRs, with careful profiling

Reply:
This performance improvement will be taken care of in another PR.